evm/src/generation/prover_input.rs

```diff
@@ -229,6 +240,175 @@ impl<F: Field> GenerationState<F> {
+        // the previous 32 bytes in the code (including opcodes and pushed bytes)
+        // are PUSHXX and the address is in its range
+
+        let proof = CodeIterator::until(&code, self.last_jumpdest_address + 1).fold(
+            0,
+            |acc, (pos, opcode)| {
+                let has_prefix = if let Some(prefix_start) = pos.checked_sub(32) {
+                    code[prefix_start..pos].iter().enumerate().fold(
+                        true,
+                        |acc, (prefix_pos, &byte)| {
+                            acc && (byte > get_push_opcode(32)
+                                || (prefix_start + prefix_pos) as i32
+                                    + (byte as i32 - get_push_opcode(1) as i32)
+                                    + 1
+                                    < pos as i32)
+                        },
+                    )
+                } else {
+                    false
+                };
+                if has_prefix {
+                    pos - 32
+                } else {
+                    acc
+                }
+            },
+        );
+        Ok(proof.into())
+    }
+}
+
+impl<F: Field> GenerationState<F> {
+    fn generate_jumpdest_table(&mut self) -> Result<(), ProgramError> {
+        let mut state = self.soft_clone();
+        let code_len = u256_to_usize(self.memory.get(MemoryAddress {
+            context: self.registers.context,
+            segment: Segment::ContextMetadata as usize,
+            virt: ContextMetadata::CodeSize as usize,
+        }))?;
+        // Generate the jumpdest table
+        let code = (0..code_len)
+            .map(|i| {
+                u256_to_u8(self.memory.get(MemoryAddress {
+                    context: self.registers.context,
+                    segment: Segment::Code as usize,
+                    virt: i,
+                }))
+            })
+            .collect::<Result<Vec<u8>, _>>()?;
+
+        // We need to set the the simulated jumpdest bits to one as otherwise
+        // the simulation will fail
+        let mut jumpdest_table = vec![];
+        for (pos, opcode) in CodeIterator::new(&code) {
+            jumpdest_table.push((pos, opcode == get_opcode("JUMPDEST")));
+            if opcode == get_opcode("JUMPDEST") {
+                state.memory.set(
+                    MemoryAddress {
+                        context: state.registers.context,
+                        segment: Segment::JumpdestBits as usize,
+                        virt: pos,
+                    },
+                    U256::one(),
+                );
+            }
+        }
+
+        // Simulate the user's code and (unnecessarily) part of the kernel code, skipping the validate table call
+        self.jumpdest_addresses = simulate_cpu_between_labels_and_get_user_jumps(
+            "validate_jumpdest_table_end",
+            "terminate_common",
+            &mut state,
+        )
+        .ok();
+
+        Ok(())
+    }
+}
```

Comment:
nit, it'd be cleaner to have `simulate_cpu_between_labels_and_get_user_jumps` return a `Result<foo, ProgramError>` so that we can just do 

```suggestion
        self.jumpdest_addresses = simulate_cpu_between_labels_and_get_user_jumps(
            "validate_jumpdest_table_end",
            "terminate_common",
            &mut state,
        )
    }
}
```